src/set/api.rs

```diff
@@ -0,0 +1,706 @@
+	/// The last element is always the maximum element in the set.
+	///
+	/// ## Original
+	///
+	/// [`BTreeSet::pop_last`](alloc::collections::BTreeSet::pop_last)
+	///
+	/// ## Examples
+	///
+	/// ```rust
+	/// use bitvec::prelude::*;
+	///
+	/// let mut set = BitSet::new();
+	/// set.insert(1);
+	/// while let Some(n) = set.pop_last() {
+	///     assert_eq!(n, 1);
+	/// }
+	/// assert!(set.is_empty());
+	/// ```
+	#[inline]
+	pub fn pop_last(&mut self) -> Option<usize> {
+		let last = self.last()?;
+		self.remove(last);
+		Some(last)
+	}
+
+	/// Adds a value to the set.
+	///
+	/// Returns whether the value was newly inserted. That is:
+	///
+	/// * If the set did not previously contain the value, `true` is returned.
+	/// * If the set already contained the value, `false` is returned.
+	///
+	/// ## Original
+	///
+	/// [`BTreeSet::insert`](alloc::collections::BTreeSet::insert)
+	///
+	/// ## Examples
+	///
+	/// ```rust
+	/// use bitvec::prelude::*;
+	///
+	/// let mut set = BitSet::new();
+	///
+	/// assert!(set.insert(2));
+	/// assert!(!set.insert(2));
+	/// assert_eq!(set.len(), 1);
+	/// ```
+	#[inline]
+	pub fn insert(&mut self, value: usize) -> bool {
+		let old_len = self.inner.len();
+		if value >= old_len {
+			self.inner.resize(value + 1, false);
+			self.inner.set(value, true);
+			false
+		}
+		else {
+			!self.inner.replace(value, true)
+		}
+	}
+
+	/// If the set contains the value, removes it from the set.
+	/// Returns whether such an element was present.
+	///
+	/// ## Original
+	///
+	/// [`BTreeSet::remove`](alloc::collections::BTreeSet::remove)
+	///
+	/// ## Examples
+	///
+	/// ```rust
+	/// use bitvec::prelude::*;
+	///
+	/// let mut set = BitSet::new();
+	///
+	/// set.insert(2);
+	/// assert!(set.remove(2));
+	/// assert!(!set.remove(2));
+	/// ```
+	#[inline]
+	pub fn remove(&mut self, value: usize) -> bool {
+		let old_len = self.inner.len();
+		if value >= old_len {
+			false
+		}
+		else {
+			let ret = self.inner.replace(value, false);
+
+			// NOTE: it's unclear how this affects performance and if we should
+			// do this automatically, or require it manually only
+			self.shrink_inner();
+
+			ret
+		}
+	}
+
+	/// Retains only the elements specified by the predicate.
+	///
+	/// In other words, remove all elements `e` for which `f(e)` returns
+	/// `false`. The elements are visited in ascending order.
+	///
+	/// ## Original
+	///
+	/// [`BTreeSet::retain`](alloc::collections::BTreeSet::retain)
+	///
+	/// ## Examples
+	///
+	/// ```rust
+	/// use bitvec::prelude::*;
+	///
+	/// let mut set = BTreeSet::new();
+	/// set.insert(1);
+	/// set.insert(2);
+	/// set.insert(3);
+	/// set.insert(4);
+	/// set.insert(5);
+	/// set.insert(6);
+	/// // Keep only the even numbers.
+	/// set.retain(|k| k % 2 == 0);
+	/// assert!(set.iter().eq([2, 4, 6].into_iter()));
+	/// ```
+	#[inline]
+	pub fn retain<F>(&mut self, mut f: F)
+	where F: FnMut(usize) -> bool {
+		self.inner
+			.iter_mut()
+			.enumerate()
```

Comment:
Fully aware that this is super inefficient, but `iter_ones` doesn't have a version that would play nice with mutation.